### ISSUE TITLE
usernameless flow demo

### ISSUE THREAD

**Issue by bmd007** [Yubico/java-webauthn-server#300]
Hi
looking at the code in the demo folder, in the [start authentication method](https://github.com/Yubico/java-webauthn-server/blob/f207b279bdf740abfc0618398c92bc2cea2557e6/webauthn-server-demo/src/main/java/demo/webauthn/WebAuthnServer.java#L422), when username is `null` (indicating username less flow),  the following happens:
```
AssertionRequestWrapper request =
          new AssertionRequestWrapper(
              generateRandom(32),
              rp.startAssertion(StartAssertionOptions.builder().username(username).build())) 
```

in which username is still empty (null) and userHandle is not defined (null again). Which in turn results in the following code being executed with null all over it:
```
            .allowCredentials(
                OptionalUtil.orElseOptional(
                        startAssertionOptions.getUsername(),
                        () ->
                            startAssertionOptions
                                .getUserHandle()
                                .flatMap(credentialRepository::getUsernameForUserHandle))
                    .map(
                        un ->
                            new ArrayList<>(credentialRepository.getCredentialIdsForUsername(un))))
```

` startAssertionOptions.getUsername()` is null, `startAssertionOptions.getUserHandle()` is empty ====> finally `allowCredentials` is null.

I think the demo doesn't really support the username less flow. 
I couldn't find enough documentations explaining how to implement the flow either.

**Comment by emlun** [comment]
That's what the "Authenticate without username" button in the demo is for. :slightly_smiling_face:

**Comment by bmd007** [comment]
that button in the demo eventually calls the code stack which I pointed to. So If I got it right, there is a problem with the demo code.

**Comment by bmd007** [comment]
Thanks for the info @emlun 

I realised there are still minor details ignored in the main code regarding username less flow. 
 1- You can't just leave the allowCredentials null. The javaScript side requires a list to be present even if empty. So an empty list is required instead of null.
2- Consequently, during finishAuthentication phase, step 5, the request has allowCredentials (an empty list). The current code tries to find a matching credential between the ones in the request and the ones in the response. Well, there is no credentials in the request (hence the username less ness). So checking the size of the list is also important.

Basically the code needs to be more defensive about lists inside optionals, and their (de-)serialization.  

I will try to create a PR, let's see if I can make it.

**Comment by bmd007** [comment]
I created https://github.com/Yubico/java-webauthn-server/pull/305. Please have a look.
The code base was too complicated for me to add tests, so please feel free to edit my PR if you find it any useful.

**Comment by emlun** [comment]
There is no problem here as far as I can tell. The [`StartAssertionOptions.username(String)` method](https://github.com/Yubico/java-webauthn-server/blob/e371103fd38003ae790f5113b9632db226df9f5d/webauthn-server-core/src/main/java/com/yubico/webauthn/StartAssertionOptions.java#L232) wraps `null` values in `Optional`, and `Optional.empty()` values are handled correctly in `FinishAssertionSteps`. The changes suggested in #305 do not seem necessary.

**Comment by emlun** [comment]
With that, I'll consider this resolved. Please let us know if there's more to discuss here.

**Comment by bmd007** [comment]
> With that, I'll consider this resolved. Please let us know if there's more to discuss here.

FinishAssertionStep does not handle the null. When username is missing, what will reach UI is a not an empty collection ([]). 

The fact that the collection is wrapped by optional, is a clear sign of such potential problems.

**Comment by bmd007** [comment]
I think you are not spending enough time looking at the code and actually debugging it. Nor understanding my PR.

**Comment by emlun** [comment]
Ah, my apologies, now I see what you mean: `finishAssertion` rejects the credential ID as "unrequested" if you explicitly set `StartAssertionOptions.allowCredentials` to a present but empty list. I was confused because the issue mostly mentions null values, which the library does handle correctly as far as I'm aware.

The updated test in 1576b3d0756f738e8c892855104fa4cb188faf59 confirms that `FinishAssertionSteps` incorrectly rejects the assertion when `allowCredentials` is explicitly set to an empty list. The same commit also fixes the issue. Thanks for the report!

**Comment by emlun** [comment]
> The javaScript side requires a list to be present even if empty.

Just so you know: this is true only of precisely the value `null`. The JS API permits `allowCredentials` to be omitted or set to `undefined`; these are both equivalent to setting `allowCredentials: []`. See these examples in the Chrome console:

With `allowCredentials: null`:
```js
> await navigator.credentials.get({
    publicKey: {
        challenge: crypto.getRandomValues(new Uint8Array(32)),
        allowCredentials: null,
    },
})

VM340:2 Uncaught TypeError: Failed to execute 'get' on 'CredentialsContainer': Failed to read the 'publicKey' property from 'CredentialRequestOptions': Failed to read the 'allowCredentials' property from 'PublicKeyCredentialRequestOptions': The provided value cannot be converted to a sequence.
    at <anonymous>:1:29
(anonymous) @ VM340:1
```

With `allowCredentials: undefined`:
```js
> await navigator.credentials.get({
    publicKey: {
        challenge: crypto.getRandomValues(new Uint8Array(32)),
        allowCredentials: undefined,
    },
})

PublicKeyCredential {rawId: ArrayBuffer(32), response: AuthenticatorAssertionResponse, authenticatorAttachment: 'platform', id: '1zBfDVjoA_QGE4kKKDNTEdOZryCYF9PsOtsCcm8_5bs', type: 'public-key'}
```

With `allowCredentials` omitted:
```js
> await navigator.credentials.get({
    publicKey: {
        challenge: crypto.getRandomValues(new Uint8Array(32)),
    },
})

PublicKeyCredential {rawId: ArrayBuffer(32), response: AuthenticatorAssertionResponse, authenticatorAttachment: 'platform', id: '1zBfDVjoA_QGE4kKKDNTEdOZryCYF9PsOtsCcm8_5bs', type: 'public-key'}
```

**Comment by bmd007** [comment]
No worries. 
I think there is still a confusion. 

The current code on master, when username is null, results in `allow Credentials`  becoming null. Thus results in an error in the JS side (as you described yourself).

The PRZ that I provided before,  solves this issue by making sure there will always be a `allow Credentials`.

**Comment by bmd007** [comment]
@emlun 

The screenshots below show what happens in the username less flow:

<img width="1624" alt="image" src="https://github.com/Yubico/java-webauthn-server/assets/7900715/8ca2dfad-69c2-42e1-989e-d54da516e80a">

<img width="1634" alt="image" src="https://github.com/Yubico/java-webauthn-server/assets/7900715/3c93686d-a946-43f0-a679-bb68501d248c">


You can see that when both username and userHandle are empty optionals (on the java side), the resulting json that reaches frontend (JS) contains a `allowCredentials` with null as value. Which as you mentioned, results in an error. 
My PR was making sure that `allowCredentials` is never an empty optional, but an empty list.

**Comment by bmd007** [comment]
you probably have some config somewhere in your serialiser which ignores empty optional fields during serialisation. But that's not the norm, simply assuming that is problematic. 

```
    public AssertionRequestWrapper startAuthentication() {
        var assertionExtensionInputs = AssertionExtensionInputs.builder()
                .uvm()
                .build();
        var startAssertionOptions = StartAssertionOptions.builder()
                .userVerification(UserVerificationRequirement.PREFERRED)
                .extensions(assertionExtensionInputs)
                .build();
        AssertionRequest assertionRequest = relyingParty.startAssertion(startAssertionOptions);
        var assertionRequestWrapper = new AssertionRequestWrapper(randomUUIDByteArray(), assertionRequest);
        assertRequestStorage.put(assertionRequestWrapper.getRequestId(), assertionRequestWrapper);
        return assertionRequestWrapper;
    }
```

in other word, serialising assertionRequestWrapper results in a json with `allowCredentials` filed present in it but with value `null` .

**Comment by emlun** [comment]
Ah, I see. Using an external serializer is an unsupported use case, and this issue is one of the reasons for that. The supported use, as instructed in both the ["Getting started" README](https://github.com/Yubico/java-webauthn-server#3-registration) and the [package JavaDoc](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/2.5.0/com/yubico/webauthn/package-summary.html), is to use the [`toCredentialsCreateJson()`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/data/PublicKeyCredentialCreationOptions.html#toCredentialsCreateJson()) and [`toCredentialsGetJson()`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/AssertionRequest.html#toCredentialsGetJson()) methods to convert the request objects to JSON.

**Comment by bmd007** [comment]
> Ah, I see. Using an external serializer is an unsupported use case, and this issue is one of the reasons for that. The supported use, as instructed in both the ["Getting started" README](https://github.com/Yubico/java-webauthn-server#3-registration) and the [package JavaDoc](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/2.5.0/com/yubico/webauthn/package-summary.html), is to use the [`toCredentialsCreateJson()`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/data/PublicKeyCredentialCreationOptions.html#toCredentialsCreateJson()) and [`toCredentialsGetJson()`](https://developers.yubico.com/java-webauthn-server/JavaDoc/webauthn-server-core/latest/com/yubico/webauthn/AssertionRequest.html#toCredentialsGetJson()) methods to convert the request objects to JSON.

Mixing (de)serialuzation logic with domain/dto classes to avoid "such" cases instead of writing a more defensive code, not gonna fly much.
No need for it either.
